pkg/manifestcomparators/comp_no_data_type_change.go

```diff
@@ -0,0 +1,90 @@
+package manifestcomparators
+
+import (
+	"fmt"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+type noDataTypeChange struct{}
+
+func NoDataTypeChange() CRDComparator {
+	return noDataTypeChange{}
+}
+
+func (noDataTypeChange) Name() string {
+	return "NoDataTypeChange"
+}
+
+func (noDataTypeChange) WhyItMatters() string {
+	return "If the data type of fields are changed, then clients that rely on those fields will not be able to read them or write them."
+}
+
+func getFieldsAndTypes(version *apiextensionsv1.CustomResourceDefinitionVersion) map[string]string {
+	fieldsAndTypes := make(map[string]string)
+	SchemaHas(version.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+		func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
+			fieldsAndTypes[simpleLocation.String()] = s.Type
+			return false
+		})
+
+	return fieldsAndTypes
+}
+
+type TypeChange struct {
+	ExistingType string
+	NewType      string
+}
```

Comment:
Do we also need to care about transitions of formats as well as types? Eg, if I were to change a `string` field to `metav1.Time`, then the `type: string` is still present inside the OpenAPI schema, but I would now also have `format: date-time` right? IIRC, in Go at least, the `metav1.Time` has its own JSON unmarshalling logic, which would then start to break if it tried to unmarshal 

I'm not actually sure what the change would do to existing data in the CRD schema, it's likely it wouldn't pass the openapi schema validation, which may or may not ratchet? I suspect that doesn't ratchet

So perhaps adding `format` to an existing `string`, or changing it, should also be considered to be an issue?

Reply:
Should format be a distinct check? 
 
For reference, in https://github.com/everettraven/crd-diff I try to make each property-based validation it's own distinct check. crd-diff already has one for types, but I also identified `format` being one that I should add in https://github.com/everettraven/crd-diff/issues/3

Reply:
That could be an acceptable outcome of this thread to be fair, it was just something I was thinking of as going through

Would there be a reason to enable data type checking, but not format checking?

Reply:
>Would there be a reason to enable data type checking, but not format checking?

I don't think so, but my only point was that maybe they shouldn't be so tightly coupled to the same validator since they are evaluating two different properties.

Reply:
What we have here is useful, so we can get it merged and follow up with either an extension to this check, or a separate one

Reply:
We could do some testing to find out what happens. That check could be done in another PR.